Save parser positions ast

mstemm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

<!--
Please remove the leading whitespace before the `/version <>` you uncommented.
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, you have to do nothing.
If yes, a release note is required:
Delete `NONE` and enter your extended release note in the block below.
Please note, the release note follows the "conventional commit specification" (https://www.conventionalcommits.org/en/v1.0.0/):
For example: `fix: broken link`.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
Save condition parser positions in ASTs and make them available to downstream tools like the filter compiler.
```

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/560#" title="Author self-approved">mstemm</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [mstemm]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

jasondellaluce commented on userspace/libsinsp/filter/ast.h:
```diff
@@ -117,9 +145,15 @@ struct SINSP_PUBLIC string_visitor: public expr_visitor
 */
 struct SINSP_PUBLIC expr
 {
+    expr();
```
What are these two constructors used for? I think we should leave this interface as abstract as possible.

Due to the same reasoning, Isn't it redundant to have all three get_pos, set_pos, and m_pos? I think we should aim to either:
- Have only `m_pos`
- Have only `get_pos` and `set_pos` and leave each implementation store their class members
- Keep all three, but turn `expr` in a class and set `m_pos` as private and all the rest as public

mstemm replied:
I'll get rid of the second constructor, I was using it but once I switched to using the static ::create() methods I didn't need it. About defining a default constructor, this is more of a style thing--some devs always want constructors defined unless the class is truly abstract, and I tend to follow that practice. But I can remove it since it doesn't do anything. (Someday defining a style guide would be useful). I did add default values to the pos_info idx/line/col though.

If the intent is that expr is an interface/fully abstract class, I can push down management of the position into the sub-classes. But if we want to keep the position as a part of the base class, I wouldn't have the derived classes access a protected member. It depends on who you ask, but many devs as well as sonarcloud (https://rules.sonarsource.com/cpp/RSPEC-3656) frown on it.

So I'd prefer the third option. I'll do that.

jasondellaluce commented on userspace/libsinsp/filter/ast.h:
```diff
@@ -117,9 +145,15 @@ struct SINSP_PUBLIC string_visitor: public expr_visitor
     virtual void accept(expr_visitor*) = 0;
     virtual bool is_equal(const expr* other) const = 0;
+
+    const pos_info& get_pos();
```
```suggestion
    const pos_info& get_pos() const;
```
If you decide to keep it.

mstemm replied:
Sure, added.

jasondellaluce commented on userspace/libsinsp/filter.h:
```diff
@@ -105,6 +105,8 @@ class SINSP_PUBLIC sinsp_filter_compiler:
 	*/
 	sinsp_filter* compile();
 
+	const libsinsp::filter::ast::pos_info& get_pos();
```
```suggestion
	const libsinsp::filter::ast::pos_info& get_pos() const;
```

mstemm replied:
Yep, added.

jasondellaluce commented on userspace/libsinsp/filter/ast.cpp:
```diff
@@ -171,6 +171,25 @@ const std::string& string_visitor::as_string()
+{
+}
+
+const libsinsp::filter::ast::pos_info& libsinsp::filter::ast::expr::get_pos()
```
Nit: we might as well inline `get_pos` and `set_pos` in ast.h right away.

mstemm replied:
This is another thing where a style guide would be useful--some devs want only declarations in header files. We don't have good rules on explicit inlining either. But if we want to allow trivial or nearly-trivial declarations, sure. I'll move them. (I'm going to move expr::get_pos()/expr::set_pos() as well then).

jasondellaluce reviewed (commented):
This is super useful! Left few suggestions. On top of those, I'd suggest adding few unit tests in `filter_parser.ut.cpp` as well so that we protect ourselves against future regressions!

Molter73 commented on userspace/libsinsp/filter/ast.h:
```diff
@@ -225,9 +265,12 @@ struct SINSP_PUBLIC not_expr: expr
+					    const libsinsp::filter::ast::pos_info& pos = s_initial_pos)
     {
-        return std::unique_ptr<not_expr>(new not_expr(std::move(c)));
+        std::unique_ptr<not_expr> ret(new not_expr(std::move(c)));
+	ret->set_pos(pos);
+	return ret;
```
This is an ultra nitpick and it should be addressed when we enforce code formatting, but I see some changes mix up tabs and space for indentations, would you mind unifying them into spaces?

mstemm replied:
We're kind of all over the map wrt spaces vs tabs though. The majority of the libsinsp code is tabs. Even many of the other files below libsinsp/filter are using tabs.

I'd rather we tackle formatting stuff like this in one change by adding a formatter as a commit hook (or whatever else is best practice).

Molter73 reviewed (commented):
The changes look good to me, other than Jason's comments I left a small nitpick on indentation formatting that would not prevent the PR from being merged.

mstemm commented:
I believe I addressed all feedback. I'm writing tests now, and I think they're revealing some off-by-one errors in the original parser code. Still taking a look to see if the bugs are real or problems in the tests.

jasondellaluce reviewed (commented):
Changes LGTM. Waiting for your feedback on testing.

mstemm commented:
Ok, added some tests. https://github.com/falcosecurity/libs/pull/560/commits/922ef97d1ac2596b6517bc31a119bce3fe716eb0 fixes one very minor bug related to parser positions--when parsing values, the position should be set after skipping any leading whitespace.

PTAL!

jasondellaluce reviewed (approved):
LGTM, thanks!

poiana commented:
LGTM label has been added.  <details>Git tree hash: b32451c7e3fad4f9729e4f07ad29084f2df6f7e0</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/560#pullrequestreview-1100376459" title="Approved">jasondellaluce</a>*, *<a href="https://github.com/falcosecurity/libs/pull/560#" title="Author self-approved">mstemm</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [mstemm]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

VadimZy commented on userspace/libsinsp/filter/ast.h:
```diff
@@ -36,6 +36,34 @@ struct list_expr;
+	uint32_t col = 1;
+};
+
+static pos_info s_initial_pos;
```
it must be const and not static

VadimZy commented on userspace/libsinsp/filter/ast.h:
```diff
@@ -36,6 +36,34 @@ struct list_expr;
+*/
+struct pos_info
+{
+	inline void reset()
```
why do we need this?

VadimZy commented on userspace/libsinsp/filter/ast.h:
```diff
@@ -36,6 +36,34 @@ struct list_expr;
+			+ ", column " + std::to_string(col);
+	}
+
+	uint32_t idx = 0;
```
why `uint32_t`? why not `size_t`

VadimZy commented on userspace/libsinsp/filter/ast.h:
```diff
@@ -200,9 +238,12 @@ struct SINSP_PUBLIC or_expr: expr
 
-    static std::unique_ptr<or_expr> create(std::vector<std::unique_ptr<expr>> &c)
+    static std::unique_ptr<or_expr> create(std::vector<std::unique_ptr<expr>> &c,
+					   const libsinsp::filter::ast::pos_info& pos = s_initial_pos)
```
const libsinsp::filter::ast::pos_info& pos = {}

VadimZy commented on userspace/libsinsp/filter/ast.h:
```diff
@@ -162,9 +197,12 @@ struct SINSP_PUBLIC and_expr: expr
 
-    static std::unique_ptr<and_expr> create(std::vector<std::unique_ptr<expr>> &c)
+    static std::unique_ptr<and_expr> create(std::vector<std::unique_ptr<expr>> &c,
+					    const libsinsp::filter::ast::pos_info &pos = s_initial_pos)
```
I don't understand why you use the short type `pos_info` a few lines above and the fully qualified type here.
are we out of the `libsinsp::filter::ast::` namespaces?

FedeDP reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/560#pullrequestreview-1105189146" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/560#pullrequestreview-1100376459" title="Approved">jasondellaluce</a>*, *<a href="https://github.com/falcosecurity/libs/pull/560#" title="Author self-approved">mstemm</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP,mstemm]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
